### PR TITLE
Add role to provision resources related to publishing egress IPs

### PR DESCRIPTION
**Pull request by dav3r**
#  #

## 🗣 Description ##

This PR creates a role that includes the permissions necessary to provision all resources related to publishing egress IPs in the DNS account.

## 💭 Motivation and context ##

This `ProvisionPublishEgressIP` role will be used by the forthcoming `cisagov/publish-egress-ip-terraform` repo, which will deploy all of the resources related to the egress IP publishing process in the DNS account.

This is part of the work for:
* https://github.com/cisagov/cool-system/issues/208
* https://github.com/cisagov/cyhy_amis/issues/284

## 🧪 Testing ##

I successfully applied this code in the DNS account and I was also able to use the `ProvisionPublishEgressIP` role to successfully deploy test versions of all of the resources related to the egress IP publishing process.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

- [ ] If any material changes were made during the PR approval process, apply them in the DNS account.

### CODE REVIEW

File: dns/assume_role_policy_doc.tf [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# ------------------------------------------------------------------------------
+# Create an IAM policy document that allows the users account to
+# assume this role.
```
Comment (mcdonnnj) [inline dns/assume_role_policy_doc.tf]: There is nothing indicating what role is being talked about in this context.

```suggestion
# assume a role.
```

File: dns/outputs.tf [hunk 2]
```diff
@@ -8,6 +8,11 @@ output "provisionaccount_role" {
   description = "The IAM role that allows sufficient permissions to provision all AWS resources in the DNS account."
 }
 
+output "provisionpublishegressip_role" {
+  value       = aws_iam_role.provisionpublishegressip_role
+  description = "The IAM role that allows sufficient permissions to provision all resources related to the publish-egress-ip Lambda in the DNS account."
+}
+
```
Comment (jsf9k) [inline dns/outputs.tf]: Please properly alphabetize this additional output.

**Review by jsf9k — CHANGES_REQUESTED** [review]
I have one small quibble.

File: dns/outputs.tf (hunk 2 above)
Comment (dav3r) [reply dns/outputs.tf]: Excellent catch- rectified in https://github.com/cisagov/cool-accounts/pull/127/commits/d2de7cdf53f981720d62dcd11299d997c9a7d785.

Comment (jsf9k) [reply dns/outputs.tf]: Your change is also nice, but I was actually referring to the fact that `primary_delegation_set` should come before `provisionaccount_role` and `provisionpublishegressip_role`.

Comment (dav3r) [reply dns/outputs.tf]: Haha, I totally missed that before, thanks. 😄  See https://github.com/cisagov/cool-accounts/pull/127/commits/69365712af3d7198735f4f6e3d916f20662dd485.

**Review by jsf9k — APPROVED** [review]

File: dns/provisionpublishegressip_policy.tf [hunk 3]
```diff
@@ -0,0 +1,180 @@
+# ------------------------------------------------------------------------------
+# Create the IAM policy that allows provisioning all resources related to
+# the publish-egress-ip Lambda in the DNS account.
+# ------------------------------------------------------------------------------
+
+data "aws_iam_policy_document" "provisionpublishegressip_doc" {
```
Comment (mcdonnnj) [inline dns/provisionpublishegressip_policy.tf:7]: Did you test that these permissions are sufficient to tear down the related resources?

File: dns/provisionpublishegressip_role.tf [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# ------------------------------------------------------------------------------
+# Create the IAM role that allows provisioning all resources related to
+# the publish-egress-ip Lambda in the DNS account.
```
Comment (mcdonnnj) [inline dns/provisionpublishegressip_role.tf]: This aligns with language used in other COOL projects.

```suggestion
# Create the IAM role that allows all of the permissions necessary to
# provision all resources related to the publish-egress-ip Lambda in
# the DNS account.
```

File: dns/provisionpublishegressip_policy.tf [hunk 5]
```diff
@@ -0,0 +1,180 @@
+# ------------------------------------------------------------------------------
+# Create the IAM policy that allows provisioning all resources related to
+# the publish-egress-ip Lambda in the DNS account.
```
Comment (mcdonnnj) [inline dns/provisionpublishegressip_policy.tf]: This aligns with language used in other COOL projects.

```suggestion
# Create the IAM policy that allows all of the permissions necessary to
# provision all resources related to the publish-egress-ip Lambda in
# the DNS account.
```

**Review by mcdonnnj — APPROVED** [review]
This makes sense. I just have some documentation suggestions and one question for your consideration.

File: dns/provisionpublishegressip_policy.tf (hunk 3 above)
Comment (dav3r) [reply dns/provisionpublishegressip_policy.tf:7]: I had tested this previously, but I wasn't able to delete the versioned bucket with the lambda-at-edge security headers jazz.  I finally had a chance to circle back to this and found the missing permission, which I added with https://github.com/cisagov/cool-accounts/pull/127/commits/685c25c4e7e6067de603eb1b9a5e7834506ec754.  Now everything can be cleanly torn down. 👍

**Pull request merged**